node_settled_handler.go

```diff
@@ -115,51 +110,61 @@ func (p *NodeSettledHandler) subscribeEvents(ctx context.Context) error {
 		if !ok {
 			continue
 		}
-		if !finalEvent.FinalHtlcEvent.Settled {
-			// TODO: Handle final fail?
-
-			continue
-		}
 
 		key := types.CircuitKey{
 			ChanID: event.IncomingChannelId,
 			HtlcID: event.IncomingHtlcId,
 		}
 
-		if err := p.handleFinalHtlc(ctx, key); err != nil {
+		err = p.handleFinalHtlc(
+			ctx, key, finalEvent.FinalHtlcEvent.Settled,
+		)
+		if err != nil {
 			return err
 		}
 	}
 }
 
 func (p *NodeSettledHandler) handleFinalHtlc(ctx context.Context,
-	key types.CircuitKey) error {
+	key types.CircuitKey, settled bool) error {
+
+	var status types.HtlcStatus
+	if settled {
+		status = types.HtlcStatusSettled
+	} else {
+		status = types.HtlcStatusFailed
+	}
 
 	htlcKey := types.HtlcKey{
 		ChanID: key.ChanID,
 		HtlcID: key.HtlcID,
 		Node:   p.lnd.PubKey(),
 	}
 
-	settledHash, err := p.persister.MarkHtlcSettled(ctx, htlcKey)
+	invoiceHash, err := p.persister.MarkHtlcFinal(ctx, htlcKey, settled)
 	switch {
-	case err == persistence.ErrHtlcAlreadySettled:
+	case err == persistence.ErrHtlcAlreadyFinal:
 		return nil
 
-	// If the htlc is not found, the final resolution was for an htlc that isn't
-	// managed by lnmux.
+	// If the htlc is not found, the final resolution was for an htlc that
+	// isn't managed by lnmux.
 	case err == types.ErrHtlcNotFound:
 		return nil
 
 	case err != nil:
-		return fmt.Errorf("unable to mark htlc %v settled: %w", key, err)
+		return fmt.Errorf("unable to mark htlc %v as %s : %w",
+			key, status, err)
 	}
 
-	p.logger.Infow("Htlc final settled received",
-		"chanID", key.ChanID, "htlcID", key.HtlcID, "hash", settledHash)
+	p.logger.Infow("Htlc final resolution received",
+		"chanID", key.ChanID,
+		"htlcID", key.HtlcID,
+		"hash", invoiceHash,
+		"status", status,
+	)
 
-	if settledHash != nil {
-		p.finalCallback(*settledHash, true)
+	if invoiceHash != nil {
+		p.finalCallback(*invoiceHash, settled)
```

Comment:
Do you want to add test coverage for the invoice failed case?

Reply:
very very weird

I did the tests:

https://github.com/bottlepay/lnmux/commit/2abdd53414e07d4ca53f850c91a24b2daab6b41d#diff-b39fc442f543ad6df386d0772fdbafafee01095292545626a39251403bec6b5cR458

But for some reason, when doing a rebase, I removed them...

Reply:
Oh oh.... maybe something else changed. You can diff to some revision from `git reflog`

Reply:
I triple checked, only this test was "lost"!

I just added the test again